### PR TITLE
reactive composable data graphics for the metrics explore line primitives (and more)

### LINKED ISSUE #460: Metrics Explorer: Implement Line Chart components
**Issue**
A clear and concise description of the problem statement.

**Importance**
An articulation of the value we expect this feature to add for our users and the business.

**Definition of Done**
An actionable description of what the application will look at feel like when this work is complete.

**Screenshots**
If applicable, add screenshots to help explain your problem and solution.

**Additional Context**

### PR DESCRIPTION
**Pull request by hamilton**
closes #460 

The goal of this PR is to provide components for the line charts in the metrics explorer. To get there, however, I figured it'd be smart to make totally reactive composable data graphic primitives. We keep punting on doing our charting in a principled way. 

With this in mind, this is a big foundational PR. It minimally touches the existing code, but rather adds new components (hence the large & of files).

The main premise: if the data or plot config values change, the entire system reactively updates. This is true for any nested composition of data graphic components, e.g. small multiples, embedded charts within charts, etc.

dm-reactive-viz

There are a lot of components here. Here's how I'd review it:

- **organization** – check `src/lib/components/data-graphic` to see how I'm breaking things out.
- **API** – check to see how the examples _look_ from an API point of view. I"m sure ther are some convenience components that can be created to make things even easier, but I think we're off to a good start with this PR. Let me know what you think.
- **examples** – go to `/dev/data-graphic/` and take a look at the code examples there.

### CODE REVIEW

File: src/lib/components/data-graphic/marks/Line.svelte [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  const xMinStore = getContext(contexts.min("x")) as ExtremumResolutionStore;
+  const xMaxStore = getContext(contexts.max("x")) as ExtremumResolutionStore;
+  const yMinStore = getContext(contexts.min("y")) as ExtremumResolutionStore;
+  const yMaxStore = getContext(contexts.max("y")) as ExtremumResolutionStore;
+
+  // get extents
+  $: [xMinValue, xMaxValue] = extent(data, (d) => d[xAccessor]);
+  $: [yMinValue, yMaxValue] = extent(data, (d) => d[yAccessor]);
+  // set your extrema here
+  $: xMinStore.setWithKey(markID, xMin || xMinValue);
```
Comment (djbarnwal) [inline src/lib/components/data-graphic/marks/Line.svelte:43]: Would it make sense to have data exposed at the top level and not using the marks? Probably through a pluggable `Data` component or through `GraphicContext`? Plot components `Axis`, `Grid` etc. all rely on the same data. I am not sure if marks should be central authority over how data is being handled in all other components.

Having data more centrally would enable us to make multiple marks on a shared plot as well. So for something like an axis running from say 0-100 we can have one line from 0-40 and a completely different line running from 30-80 with shared axis, grid etc.

File: src/routes/dev/data-graphic/components.svelte [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+          {width}
+          {height}
+          left={leftRight}
+          right={leftRight}
+          top={topBottom}
+          bottom={topBottom}
+          xMin={0}
+          xMax={100}
+          yMin={0}
+          yMax={100}
```
Comment (djbarnwal) [inline src/routes/dev/data-graphic/components.svelte:32]: Circling back to one of the previous comment made on `Line.svelte`, I think these values should be automatically set by the data itself without the need for marks

File: src/routes/dev/data-graphic/basic.svelte [hunk 3]
```diff
@@ -0,0 +1,49 @@
+<script lang="ts">
+  import {
```
Comment (djbarnwal) [inline src/routes/dev/data-graphic/basic.svelte:2]: Not sure what's happening here but I just see this when I mount this component on an empty page

Screenshot 2022-07-06 at 8 08 16 PM

**Review by djbarnwal — COMMENTED** [review]
Haven't gone through the whole thing in detail yet but I am adding some initial thoughts on the API around the data graphics. Also these comments are not exactly reviews/requested changes but just discussion points which I thought were interesting and came to my mind while browsing through the codebase.

File: src/lib/components/data-graphic/functional-components/WithSimpleLinearScale.svelte [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<!--
+@component
+Takes in a domain and range, and outputs in a slot the derived linear scale
+-->
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/lib/components/data-graphic/functional-components/WithSimpleLinearScale.svelte:4]: That's a smart idea! I could see that being the basis for how we might also handle things like stacking, etc.

File: src/routes/dev/data-graphic/metrics.svelte [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    top={8}
+    bottom={4}
+    yMin={0}
+    bodyBuffer={0}
+    xType="date"
+    yType="number"
+  >
+    <div />
+    <GraphicContext top={24} height={24} let:config>
+      <svg width={config.width} height={24}>
```
[the remark this answers is not in the document]
Comment (hamilton) [reply src/routes/dev/data-graphic/metrics.svelte:136]: agreed, most consumption should use `SimpleDataGraphic`. Good catch!

File: src/lib/components/data-graphic/marks/Line.svelte (hunk 1 above)
Comment (hamilton) [reply src/lib/components/data-graphic/marks/Line.svelte:43]: I agree ~ I think we need to have it both ways, both as a top-level thing, and at the mark level. It's not always the case that the user  is operating with a single coherent data frame. The flexibility of being able to put data at either the shape or the plot level is pretty useful, for instance w/ `ggplot` you can put data at the top `ggplot` component and / or any of the marks, and the graphics in R will reconcile these plot bounds. Similar for us.

File: src/routes/dev/data-graphic/components.svelte (hunk 2 above)
Comment (hamilton) [reply src/routes/dev/data-graphic/components.svelte:32]: agreed, though it's pretty common to need to set the min and max values manually if the plot bounds are not actually conveying what's needed.

**Pull request merged**